torchelastic/distributed/launch.py

```diff
@@ -0,0 +1,369 @@
+
+9. `rdzv_backend` - the backend store of rendezvous (e.g. etcd). This is
+typically a strongly consistent key-value store.
+
+10. `rdzv_endpoint` - rdzv backend server endpoint in `host:port` format.
+
+A `Node` runs `LOCAL_WORLD_SIZE` workers which comprise a `LocalWorkerGroup`.
+The union of all `LocalWorkerGroups` in the nodes in the job comprise the
+`WorkerGroup`.
+
+**Deployment:**
+
+This launcher starts on agent process which creates and monitors
+a local worker group. For multi-node distributed training, simply
+start the launcher with the same arguments on all the nodes participating
+in training.
+
+When using a job/cluster manager the entry point command to the multi-node
+job is invoking this launcher.
+
+**Failure Modes:**
+
+1. Worker failure - For a training job with `n` workers, if `k < n` workers fail
+all workers are stopped and restarted up to `max_restarts`.
+
+2. Agent failure - An agent failure results in local worker group failure,
+it is up to the job manager to fail the entire job (gang semantics) or attempt
+to replace the node. Both behaviors are supported by the agent.
+
+3. Node failure - Same as agent failure.
+
+**Membership Changes:**
+
+1. Node departure (scale-down) - agent is notified of the departure,
+all existing workers are stopped, a new `Worker Group` is formed and all
+workers are started with a new `RANK` and `WORLD_SIZE`.
+
+2. Node arrival (scale-up) - the new node is admitted to the job,
+all existing workers are stopped, a new `Worker Group` is formed and all
+workers are started with a new `RANK` and `WORLD_SIZE`.
+
+
+**Important Notices:**
+
+1. All the important notices section of `torch.distributed.launch`
+ apply to this module as well
+
+2. The environment variables necessary to initialize a torch process group
+are provided to you by this module. To initialize a process group simply run
+
+::
+    >>> torch.distributed.init_process_group(backend="gloo|nccl",
+                                             init_method="env://")
+
+3. On failures or membership changes your surviving workers are killed
+immediately. Make sure to checkpoint your progress. The frequency of
+checkpoints should be proportional to the amount of lost work you can
+tolerate.
+
+4. This module only supports homogenous `LOCAL_WORLD_SIZE`. That is,
+it is assumed that all nodes run the same number of local workers (per role).
+
+5. `RANK` is not stable. Between restarts, the local workers on a node
+can be assgined a different range of ranks than before. NEVER hard code
+any assumptions about the stable-ness of ranks or some correlation between
+`RANK` and `LOCAL_RANK`.
+
+6. When using elasticity (`min_size != max_size`) DO NOT hard code
+assumptions about `WORLD_SIZE` as the world size can change due as
+nodes are allowed to leave and join.
+
+7. It is recommended your script have the following structure
+
+::
+    pre_initialize()
+    load_checkpoint(checkpoint_path)
+    initialize()
+    start_train()
```

Comment:
would I be correct if I assume this is just a recommendation and it launch script only requires the user program to be either an executable or a Python script?

Reply:
yes this is just a recommendation, but with the caveat that with no checkpoints your program will start from the beginning when there are faults or scaling events.